### PR TITLE
Installation assistant Wazuh API default password change

### PR DESCRIPTION
**Pull request by miguelfdez99**
|Related issue|
|---|
|#1523|

## Description

The unattended installation did not change the default password for wazuh and wazuh-wui users for the API. It was necessary to implement it.
- The new API password must contain a special character.

## Logs example

- New passwords added to passwords.wazuh file

-  wazuh-wui password in file /usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml

## Tests

- Distributed
https://devel.ci.wazuh.info/job/Test_unattended_distributed/464/

- All in one
https://ci.wazuh.info/job/Test_unattended/78/
https://ci.wazuh.info/job/Test_unattended/77/

### CODE REVIEW

File: unattended_installer/install_functions/installCommon.sh [hunk 1]
```diff
@@ -528,3 +528,29 @@ function installCommon_startService() {
     fi
 
 }
+
+function installCommon_changePasswordAPI() {
+
+    password_wazuh=$(cat /tmp/wazuh-install-files/passwords.wazuh | awk -F': ' '{print $2}' | tail -n 6 | head -n 1)
```
Comment (DFolchA) [inline unattended_installer/install_functions/installCommon.sh:534]: What happens if the user has set the passwords in a different order?

File: unattended_installer/install_functions/installCommon.sh [hunk 2]
```diff
@@ -528,3 +528,29 @@ function installCommon_startService() {
...
+
+    TOKEN_WUI=$(curl -s -u wazuh-wui:wazuh-wui -k -X GET "https://localhost:55000/security/user/authenticate?raw=true")
+    eval 'curl -s -k -X PUT -H "Authorization: Bearer $TOKEN_WUI" -H "Content-Type: application/json" -d "$WAZUH_WUI_PASS" "https://localhost:55000/security/users/2" -o /dev/null'
+
+}
+
+function installCommon_updateDashborad_WUI_Password() {
+
+    if [ -f "/usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml" ]; then
+        password_wazuh_wui=$(cat /tmp/wazuh-install-files/passwords.wazuh | awk -F': ' '{print $2}' | tail -n 2 | head -n 1)
```
Comment (DFolchA) [inline unattended_installer/install_functions/installCommon.sh:550]: Same, what if the user sets the passwords in a different order?

File: unattended_installer/install_functions/installCommon.sh [hunk 3]
```diff
@@ -528,3 +528,29 @@ function installCommon_startService() {
     fi
 
 }
+
+function installCommon_changePasswordAPI() {
```
Comment (DFolchA) [inline unattended_installer/install_functions/installCommon.sh:532]: This function should be included in the password tool.

File: unattended_installer/install_functions/installCommon.sh [hunk 4]
```diff
@@ -528,3 +528,29 @@ function installCommon_startService() {
...
+
+    TOKEN=$(curl -s -u wazuh:wazuh -k -X GET "https://localhost:55000/security/user/authenticate?raw=true")
+    eval 'curl -s -k -X PUT -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json" -d "$WAZUH_PASS" "https://localhost:55000/security/users/1" -o /dev/null'
+
+    TOKEN_WUI=$(curl -s -u wazuh-wui:wazuh-wui -k -X GET "https://localhost:55000/security/user/authenticate?raw=true")
+    eval 'curl -s -k -X PUT -H "Authorization: Bearer $TOKEN_WUI" -H "Content-Type: application/json" -d "$WAZUH_WUI_PASS" "https://localhost:55000/security/users/2" -o /dev/null'
+
+}
+
+function installCommon_updateDashborad_WUI_Password() {
```
Comment (DFolchA) [inline unattended_installer/install_functions/installCommon.sh:547]: This function should be included in the password tool.

**Review by DFolchA — CHANGES_REQUESTED** [review]

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 5]
```diff
@@ -380,3 +381,62 @@ function passwords_runSecurityAdmin() {
...
+    echo "  password: $password_wazuh" >> "${gen_file}"
+    echo ""	>> "${gen_file}"
+    echo "# New password for wazuh-wui API" >> "${gen_file}"
+    echo "  username: wazuh_wui" >> "${gen_file}"
+    echo "  password: $password_wazuh_wui" >> "${gen_file}"
+    echo ""	>> "${gen_file}"
+
+}
+
+function passwords_changePasswordAPI() {
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:418]: We need to add an option to change the API password.

**Review by DFolchA — CHANGES_REQUESTED** [review]

**Review by rauldpm — CHANGES_REQUESTED** [review]

File: unattended_installer/passwords_tool/passwordsMain.sh [hunk 6]
```diff
@@ -19,6 +19,12 @@ function getHelp() {
     echo -e "        -a,  --change-all"
     echo -e "                Changes all the Wazuh indexer user passwords and prints them on screen."
     echo -e ""
+    echo -e "        -ai,  --api-id <currentPassword> <id>"
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsMain.sh:22]: Avoid options with more than one variable.

**Review by DFolchA — CHANGES_REQUESTED** [review]

**Review by DFolchA — CHANGES_REQUESTED** [review]

**Comment by miguelfdez99** [comment]
Jenkins builds:
 - [Unattended distributed ](https://ci.wazuh.info/job/Test_unattended_distributed/4/) ✅ 
 - All in one:
   - [RedHat 7](https://ci.wazuh.info/job/Test_unattended/91/) ✅ 
   - [RedHat 8](https://ci.wazuh.info/job/Test_unattended/89/) ✅ 
   - [Ubuntu Bionic](https://ci.wazuh.info/job/Test_unattended/86/) ✅ 
   - [Centos7](https://ci.wazuh.info/job/Test_unattended/85/) ✅

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 7]
```diff
@@ -380,3 +380,82 @@ function passwords_runSecurityAdmin() {
...
+
+    #Change API password tool
+
+    if [[ -n "${api}" ]]; then
+        if [[ -n "${adminAPI}" ]]; then
+        common_logger -nl $'Changing API user '${nuser}' password'
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:423]: Use the same notation as the rest of the script:
```
"string ${var} "
```

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 8]
```diff
@@ -380,3 +380,82 @@ function passwords_runSecurityAdmin() {
...
+        WAZUH_PASS_API='{"password":"'"$password"'"}'
+        TOKEN_API=$(curl -s -u "${adminUser}":"${adminPassword}" -k -X GET "https://localhost:55000/security/user/authenticate?raw=true")
+        eval 'curl -s -k -X PUT -H "Authorization: Bearer $TOKEN_API" -H "Content-Type: application/json" -d "$WAZUH_PASS_API" "https://localhost:55000/security/users/${id}" -o /dev/null'
+        common_logger -nl $'API password changed'
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:427]: same

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 9]
```diff
@@ -380,3 +380,82 @@ function passwords_runSecurityAdmin() {
...
+        common_logger -nl $'The new password for user '${nuser}' is '${password}''
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:428]: same

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 10]
```diff
@@ -380,3 +380,82 @@ function passwords_runSecurityAdmin() {
...
+        else
+        common_logger -nl $'Changing API user '${nuser}' password'
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:430]: Same

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 11]
```diff
@@ -380,3 +380,82 @@ function passwords_runSecurityAdmin() {
...
+        WAZUH_PASS_API='{"password":"'"$password"'"}'
+        TOKEN_API=$(curl -s -u "${nuser}":"${currentPassword}" -k -X GET "https://localhost:55000/security/user/authenticate?raw=true")
+        eval 'curl -s -k -X PUT -H "Authorization: Bearer $TOKEN_API" -H "Content-Type: application/json" -d "$WAZUH_PASS_API" "https://localhost:55000/security/users/${id}" -o /dev/null'
+        common_logger -nl $'API password changed'
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:434]: Same

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 12]
```diff
@@ -380,3 +380,82 @@ function passwords_runSecurityAdmin() {
...
+
+}
+
+function passwords_updateDashborad_WUI_Password() {
+
+    if [ -f "/usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml" ]; then
+        password_wazuh_wui=$(< /tmp/wazuh-install-files/passwords-api.wazuh awk '$2 == "wazuh_wui" {getline;print;}' | awk -F': ' '{print $2}')
+        eval 'sed -i "s|password: wazuh-wui|password: ${password_wazuh_wui}|g" /usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml'
+    else
+        echo "ERROR: File /usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml does not exist"
```
Comment (DFolchA) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:458]: Use logger

**Review by DFolchA — CHANGES_REQUESTED** [review]

**Review by DFolchA — APPROVED** [review]
LGTM

**Comment by rauldpm** [comment]
The unattended install checks are getting stuck in the test_check_wazuh_api_status check, this causes the checkruns not to finish and there is no log to show since the connection with the aws worker dies.

It is necessary to add a time limit after which the test must be marked as failed.

Check the manual execution of said test_unattended.py tests

```
[root@ip-172-31-37-212 install]# python3 -m pytest --tb=long test_unattended.py -v -m "wazuh or wazuh_worker or indexer or dashboard"
========================================= test session starts ==========================================
platform linux -- Python 3.6.8, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /bin/python3
cachedir: .pytest_cache
rootdir: /tmp/unattended/tests/unattended/install, configfile: pytest.ini
collected 22 items / 3 deselected / 19 selected                                                        

test_unattended.py::test_check_wazuh_manager_authd PASSED                                        [  5%]
test_unattended.py::test_check_wazuh_manager_db PASSED                                           [ 10%]
test_unattended.py::test_check_wazuh_manager_execd PASSED                                        [ 15%]
test_unattended.py::test_check_wazuh_manager_analysisd PASSED                                    [ 21%]
test_unattended.py::test_check_wazuh_manager_syscheckd PASSED                                    [ 26%]
test_unattended.py::test_check_wazuh_manager_remoted PASSED                                      [ 31%]
test_unattended.py::test_check_wazuh_manager_logcollec PASSED                                    [ 36%]
test_unattended.py::test_check_wazuh_manager_monitord PASSED                                     [ 42%]
test_unattended.py::test_check_wazuh_manager_modulesd PASSED                                     [ 47%]
test_unattended.py::test_check_wazuh_manager_apid PASSED                                         [ 52%]
test_unattended.py::test_check_filebeat_process PASSED                                           [ 57%]
test_unattended.py::test_check_indexer_process PASSED                                            [ 63%]
test_unattended.py::test_check_dashboard_process PASSED                                          [ 68%]
test_unattended.py::test_check_indexer_cluster_status_not_red PASSED                             [ 73%]
test_unattended.py::test_check_dashboard_status PASSED                                           [ 78%]
test_unattended.py::test_check_wazuh_api_status 


```

File: unattended_installer/install_functions/dashboard.sh [hunk 13]
```diff
@@ -134,6 +134,8 @@ function dashboard_initialize() {
         common_logger -nl "You can access the web interface https://${nodes_dashboard_ip}\n    User: admin\n    Password: ${u_pass}"
     fi
 
+    passwords_updateDashborad_WUI_Password
```
[the remark this answers is not in the document]
Comment (DFolchA) [reply unattended_installer/install_functions/dashboard.sh]: Done [39d59da](https://github.com/wazuh/wazuh-packages/pull/1548/commits/39d59da63daacf4e3ea20cbbad770b8624917e73)

File: unattended_installer/install_functions/installCommon.sh [hunk 14]
```diff
@@ -145,6 +145,8 @@ function installCommon_createInstallFiles() {
         fi
         gen_file="/tmp/wazuh-install-files/passwords.wazuh"
         passwords_generatePasswordFile
+        api_file="/tmp/wazuh-install-files/passwords-api.wazuh"
```
[the remark this answers is not in the document]
Comment (DFolchA) [reply unattended_installer/install_functions/installCommon.sh]: Done [39d59da](https://github.com/wazuh/wazuh-packages/pull/1548/commits/39d59da63daacf4e3ea20cbbad770b8624917e73)

**Review by alberpilot — APPROVED** [review]
LGTM

**Pull request merged**